src/sphinx_antsibull_ext/directives.py

```diff
@@ -33,8 +39,75 @@ def run(self):
         return node.children
 
 
+class _Links(Directive):
+    has_content = True
+
+    def _get_string(
+        self, entry: Mapping, index: int, key: str, exp_type, optional: bool = False
+    ):
+        value = entry.get(key)
+        if value is None and optional:
+            return value
+        if isinstance(value, exp_type):
+            return value
+        raise self.error(
+            f"{index + 1}th entry in {self.name}:"
+            f" entry '{key}' must be {exp_type}, but found {value!r}"
+        )
+
+    def run(self):
+        self.assert_has_content()
+        node = nodes.bullet_list("\n".join(self.content), classes=["ansible-links"])
+        content = "\n".join(self.content)
+        try:
+            data = load_yaml_bytes(content.encode("utf-8"))
+        except Exception as exc:
+            raise self.error(
+                f"Error while parsing content of {self.name} as YAML: {exc}"
+            ) from exc
+        if not isinstance(data, Sequence):
+            raise self.error(
+                f"Content of {self.name} must be a YAML list, got {data!r} - {content!r}"
+            )
+        for index, entry in enumerate(data):
+            if not isinstance(entry, Mapping):
+                raise self.error(
+                    f"Content of {self.name} must be a YAML list consisting of mappings:"
+                    " item {index + 1} is not a mapping, got {entry!r}"
+                )
+            title = self._get_string(entry, index, "title", str)
+            url = self._get_string(entry, index, "url", str, optional=True)
+            ref = self._get_string(entry, index, "ref", str, optional=True)
+            external = (
+                self._get_string(entry, index, "external", bool, optional=True) or False
+            )
+            if (url is None) == (ref is None):
+                raise self.error(
+                    f"{index + 1}th entry in {self.name}:"
+                    " exactly one of 'url' and 'ref' must be provided"
```

Comment:
```suggestion
                    " exactly one 'url' and one 'ref' must be provided"
```

Reply:
This one should be an either-or: either 'url' is there, or 'ref' is there. Both aren't allowed, and not having any of them is also not allowed.

Reply:
For me they mean exactly the same. Could you try to explain what's the difference for you?

Reply:
I can try my best to explain it. it's kind of something that just sounds "right" to me rather than it being a formal rule that I'm aware of.

basically "one of 'url' and 'ref'" conveys that you need one of each. not that you need either one url or one ref.

Reply:
Ah, I see the problem. Would using `or` instead of `and` make it better? I'm wondering how to do this with three elements, like `Exactly one of a, b, or c must be provided`. Or can you simply write `Either a, b, or c must be provided`?

Reply:
I think the latter sounds clearer to me (`Either a, b, or c must be provided`). "One of" feels like something is missing in the construction, like `one of the following`.

Reply:
Ah, that's much better! Thanks :)